File: Week-1-April-1st-7th/Move-Zeroes/Move-Zeroes.cpp
// Method 1: Time complexity=O(n^2), Space complexity: O(1)
class Solution {
public:
    void moveZeroes(vector<int>& a) {
        for(int i = 0 ; i < a.size() - 1; i++){
            // var i only follows zero elements
            if(a[i] != 0) continue;
            // var j only follows non-zero elements
            for(int j = i + 1; j < a.size(); j++){
                if(a[j] == 0) continue;
                else{
                    int tmp = a[i];
                    a[i] = a[j];
                    a[j] = tmp;
                    break; // after swap 2 elements, we must break the loop to avoid subsequent elements with a[i]
                } 
            }
        }
    }
};

// Method 2: Time complexity; O(n), Space comlexity: O(n)
class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        vector<int> tmp;
        for(auto v : nums){
            if(v != 0){
                tmp.push_back(v);
            }
        }
        for(int i = 0 ; i < nums.size(); i++){
            if(i < tmp.size()){
                nums[i] = tmp.at(i);
            }
            else{
                nums[i] = 0;
            }
        }
    }
};

// Method 3: Time complexity=O(n), Space complexity: O(1)
class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        // the number of non-zero elements
        int non = 0;
        for(auto v : nums){
            if(v != 0){
                nums[non++] = v;
            }
        }
        for(int i = non; i < nums.size(); i++){
            nums[i] = 0;
        }
    }
};

File: Week-1-April-1st-7th/Move-Zeroes/Move-Zeroes.py
# Method 1: Time complexity=O(n^2), Space complexity: O(1)
class Solution:
    def moveZeroes(self, a: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        for i in range(len(a)-1):
            if a[i] != 0:
                continue
            for j in range(i + 1, len(a)):
                if a[j] == 0:
                    continue
                else:
                    a[i], a[j] = a[j], a[i]
                    break

# Method 2: Time complexity=O(n), Space complexity: O(n)
class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        tmp = []
        for v in nums:
            if v != 0:
                tmp.append(v)
        for i in range(len(nums)):
            if i < len(tmp):
                nums[i] = tmp[i]
            else:
                nums[i] = 0

# Method 3: Time complexity=O(n), Space complexity: O(1)
class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        # the number of non-zero elements 
        non = 0
        for v in nums:
            if v != 0:
                nums[non] = v
                non += 1
        for i in range(non, len(nums)):
            nums[i] = 0
        
        